src/NHibernate/Loader/Loader.cs

```diff
@@ -937,6 +947,17 @@ private void CheckVersion(int i, IEntityPersister persister, object id, object e
 					obj = session.GetEntityUsingInterceptor(key);
 					var alreadyLoaded = obj != null;
 					var persister = persisters[i];
+					if (IsChildFetchEntity(i))
+					{
+						if (!alreadyLoaded && mustLoadMissingEntity)
+						{
+							// Missing in session while its data has not been selected: fallback on immediate load
+							obj = session.ImmediateLoad(key.EntityName, key.Identifier);
+						}
+						rowResults[i] = obj;
+						continue;
```

Comment:
Now completely skipping `InstanceAlreadyLoaded` and `CacheByUniqueKey` for child-fetch cases.
`InstanceAlreadyLoaded` handle the locking: should we lock a thing for which we only have selected its identifier, and which the query was designed to not fetch?
Likewise, handling caching for an entity which the query is not meant to actually load should not be done.